163 Add new 'admin-framework-manager' role

katstevens opened the pull request:
Trello: https://trello.com/c/iUS9D0ll/163-add-new-framework-manager-role

An admin framework manager can:
- View application statistics
- Find suppliers by name and DUNS number (then view lists of that supplier's services and users)
- Find services by service ID (but not edit)
- View agreements (but not approve or countersign)
- Download agreement files
- Download user lists
- Upload communications sent to suppliers (already available)
- Publish clarification questions (already available)

An admin manager can:
- See users with the `admin-framework-manager` role in the admin list view
![manage-users-framework-manager](https://user-images.githubusercontent.com/3492540/33555805-9699977c-d8f9-11e7-9a2a-3492bc963355.png)

- Invite a new admin user with the `admin-framework-manager` role
![framework-manager-invite](https://user-images.githubusercontent.com/3492540/33555849-c8c4636c-d8f9-11e7-84b1-f4e8304d54a4.png)

- Existing admin users can be edited to assign/remove the `admin-framework-manager` role
![framework-manager-edit-user](https://user-images.githubusercontent.com/3492540/33555830-b699bbb0-d8f9-11e7-99a9-efc23694f562.png)

Also includes fixes for the support role being able to:
- view (but not edit) service pages
- view the list of services for a supplier (but not remove/edit)

TheDoubleK commented on tests/app/main/views/test_agreements.py:
```diff
@@ -347,3 +354,17 @@ def test_status_filtered_final_with_status_redirects_to_list(self, data_api_clie
+        ("admin-framework-manager", 302),
+        ("admin-manager", 403),
+    ])
+    def test_next_agreement_redirect_only_accessible_to_specific_user_roles(self, data_api_client, role, expected_code):
```
Hmm, thinking about this the "next agreement" should maybe only work for sourcing, as it is the redirect target after a successful "POST" to the agreements page (to approve or put on hold), so "view only" roles should never hit the "next agreement" route as far as I can think.

katstevens replied:
The 'next agreement' URL is on the `view_signed_agreement` template (i.e. as a link, not a form action)?

TheDoubleK replied:
Ah is it? OK ignore me then.

TheDoubleK commented on tests/app/main/views/test_services.py:
```diff
@@ -62,65 +75,88 @@ class TestServiceView(LoggedInApplicationTest):
-    #     assert document.xpath(
-    #         "normalize-space(string(//td[@class='summary-item-field']//*[@class='service-id']))"
-    #     ) == "1412"
+    def test_service_view_status_disabled(self, data_api_client):
```
I think these tests should have been deleted, rather than reinstated? That three-radio-button form that these are testing is old news.

katstevens replied:
Rebase problems again 😿

TheDoubleK commented on tests/app/main/views/test_suppliers.py:
```diff
@@ -33,6 +51,24 @@ def test_services_link_is_shown_to_users_with_right_roles(self, data_api_client,
         )
 
+    @pytest.mark.parametrize("role, link_should_be_visible", [
+        ("admin", False),
```
I just double-checked the permissions document and `admin` should definitely be able to "Update a supplier’s company name"

katstevens replied:
OK - will add that back in.

TheDoubleK commented on tests/app/main/views/test_users.py:
```diff
@@ -271,39 +282,51 @@ def test_get_form_with_valid_framework(self, data_api_client):
+        ("admin-framework-manager", 200),
+        ("admin-manager", 403),
+    ])
+    def test_supplier_csv_is_only_accessible_to_specific_user_roles(self, data_api_client, role, expected_code):
```
Weird how `test_download_user_lists_link_is_shown_to_users_with_the_right_role` checks that Support don't see a link to the lists, but then we test that `admin` can download the lists if they visit the URL directly.  I think it's OK for here, as I'm working on splitting the lists and will remove `admin` from being able to do so.

katstevens commented on app/main/views/suppliers.py:
```diff
@@ -90,7 +90,7 @@ def view_supplier_declaration(supplier_id, framework_slug):
 
 @main.route('/suppliers/<supplier_id>/agreements/<framework_slug>', methods=['GET'])
-@role_required('admin-ccs-sourcing')
+@role_required('admin-ccs-sourcing', 'admin-framework-manager')
```
Yep - I'd forgotten we were going to do the category permissions for this at the same time, I'm just about to add a commit for that now.

katstevens replied:
I'll do the template changes for signed agreements in a separate commit.

katstevens commented on app/templates/view_service.html:
```diff
@@ -140,13 +140,40 @@
       {% endcall %}
     {% endfor %}
+
+    {% if current_user.has_role('admin-ccs-category') %}
```
There was definitely some rebase wrangling, some tests were left in but commented out! I'll double check what's meant to be there.

katstevens commented on app/templates/_view_suppliers_edit_list.html:
```diff
@@ -9,6 +9,15 @@
   {% set field_headings = field_headings + [summary.hidden_field_heading("Services")] %}
 {% endif %}
 
+{% if current_user.has_any_role('admin', 'admin-framework-manager') %}
```
I'm not sure how the Change name got removed for Support, to be honest - I guess there must have been some miscommunication somewhere. Anyway, it's now back in 👍

katstevens commented:
@TheDoubleK I think I've addressed all the comments now. @benvand do you want to double check and make sure I've deleted the right tests after the rebase?

TheDoubleK commented:
One final thing I just noticed... for the category and support users who are view-only on agreements, should we change/remove the hint text from homepage links, which currently says eg. "Approve G-Cloud 9 agreements for countersigning"?

I'm am completely happy for this to go in as-is though and we fix that in the big homepage re-org ticket.  In fact, I'll leave a comment on that ticket now.

TheDoubleK commented:
OK so that ticket already has the link changing to a plain old "View agreements" for all users, so let's stick with this as it is.
